src/libexpr/search-path.hh

```diff
@@ -0,0 +1,101 @@
+#pragma once
+///@file
+
+#include <optional>
+
+#include "types.hh"
+#include "comparator.hh"
+
+namespace nix {
+
+/**
+ * A "search path" is a list of ways look for something, used with
+ * `builtins.findFile`.
+ */
+struct SearchPath
+{
+    /**
+     * A single element of a `SearchPath`. A key-value pair.
+     */
+    struct Elem;
+
+    /**
+     * The key of a `SearchPath::Elem` key-value pair.
+     *
+     * Called a "prefix" because it takes the form of a prefix of a file
+     * path (first `n` path components). When looking up a path, to use
+     * a `SearchPath::Elem`, its `Prefix` must match the path.
+     */
+    struct Prefix;
+
+    /**
+     * The key of a `SearchPath::Elem` key-value pair.
+     *
+     * It is either a path or a URL (with certain restrictions / extra
+     * structure).
+     *
+     * If the prefix of the patch we are looking up matches, we then see
+     * if the rest of the path points to something that exists within
+     * the directory denoted by this.
+     */
+    struct Path;
+
+    /**
+     * The list of search path elements. Each one is checked for a patch
+     * when looking up. (The actual lookup entry point is in `EvalState`
+     * not here.)
+     */
+    std::list<SearchPath::Elem> elements;
+
+    /**
+     * Parse a string into a `SearchPath`
+     */
+    static SearchPath parse(const Strings & rawElems);
+};
+
+struct SearchPath::Prefix
+{
+    /**
+     * Underlying string
+     *
+     * @todo Should we normalize this when constructing a `SearchPath::Prefix`?
+     */
+    std::string s;
+
+    GENERATE_CMP(SearchPath::Prefix, me->s);
+
+    /**
+     * If the path possibly matches this search path element, return the
+     * suffix that we should look for inside the resolved value of the
+     * element
+     */
```

Comment:
```suggestion
     * Note the double optionality in the name. While we might have a matching prefix, the suffix may not exist.
     */
```

I suppose this was intentional, and I happen to have read about it today, but it doesn't _really_ come across to me this way yet, based on just the name.